File: backend/app/services/cache.py
import hashlib
import json
from typing import Any

from app.core.config import settings
from app.db.redis import get_redis


def cache_key(namespace: str, payload: Any) -> str:
    digest = hashlib.sha256(json.dumps(payload, sort_keys=True, default=str).encode()).hexdigest()
    return f"{settings.cache_prefix}:{namespace}:{digest}"


async def get_json(key: str) -> Any | None:
    raw = await get_redis().get(key)
    return json.loads(raw) if raw else None


async def set_json(key: str, value: Any, ttl_seconds: int) -> None:
    await get_redis().setex(key, ttl_seconds, json.dumps(value, default=str))

File: backend/src/services/cache.ts
import crypto from "crypto";
import { config } from "../core/config.js";
import { getRedis } from "../db/redis.js";

export function cacheKey(namespace: string, payload: any): string {
  const serialized = JSON.stringify(payload, Object.keys(payload || {}).sort());
  const digest = crypto.createHash("sha256").update(serialized).digest("hex");
  return `${config.cachePrefix}:${namespace}:${digest}`;
}

export async function getJson<T = any>(key: string): Promise<T | null> {
  const redis = getRedis();
  if (!redis) return null;
  try {
    const raw = await redis.get(key);
    return raw ? (JSON.parse(raw) as T) : null;
  } catch {
    return null;
  }
}

export async function setJson(key: string, value: any, ttlSeconds: number): Promise<void> {
  const redis = getRedis();
  if (!redis) return;
  try {
    await redis.setex(key, ttlSeconds, JSON.stringify(value));
  } catch {
    // Gracefully handle cache write failures
  }
}
